src/append/rolling_file/policy/compound/trigger/onstartup.rs

```diff
@@ -0,0 +1,116 @@
+//!
+//! Requires the `onstartup_trigger` feature.
+
+use std::sync::Once;
+
+use crate::append::rolling_file::{policy::compound::trigger::Trigger, LogFile};
+
+#[cfg(feature = "config_parsing")]
+use crate::config::{Deserialize, Deserializers};
+
+/// Configuration for the onstartup trigger.
+#[cfg(feature = "config_parsing")]
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug, Default, serde::Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct OnStartUpTriggerConfig {
+    #[serde(default = "default_min_size")]
+    min_size: u64,
+}
+
+#[cfg(feature = "config_parsing")]
+fn default_min_size() -> u64 {
+    1
+}
+
+/// A trigger which rolls the log on startup.
+#[derive(Debug)]
+pub struct OnStartUpTrigger {
+    min_size: u64,
+    initial: Once,
+}
+
+impl OnStartUpTrigger {
+    /// Returns a new trigger which rolls the log on startup.
+    pub fn new(min_size: u64) -> OnStartUpTrigger {
+        OnStartUpTrigger {
+            min_size,
+            initial: Once::new(),
+        }
+    }
+}
+
+impl Trigger for OnStartUpTrigger {
+    fn trigger(&self, file: &LogFile) -> anyhow::Result<bool> {
+        let mut result = false;
+        self.initial.call_once(|| {
+            if file.len_estimate() >= self.min_size {
+                result = true;
+            }
+        });
+        Ok(result)
+    }
+
+    fn is_pre_process(&self) -> bool {
+        true
+    }
+}
+
+/// A deserializer for the `OnStartUpTrigger`.
+///
+/// # Configuration
+///
+/// ```yaml
+/// kind: onstartup
+///
+/// ```
+#[cfg(feature = "config_parsing")]
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug, Default)]
+pub struct OnStartUpTriggerDeserializer;
+
+#[cfg(feature = "config_parsing")]
+impl Deserialize for OnStartUpTriggerDeserializer {
+    type Trait = dyn Trigger;
+
+    type Config = OnStartUpTriggerConfig;
+
+    fn deserialize(
+        &self,
+        config: OnStartUpTriggerConfig,
+        _: &Deserializers,
+    ) -> anyhow::Result<Box<dyn Trigger>> {
+        Ok(Box::new(OnStartUpTrigger::new(config.min_size)))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn pre_process() {
+        let trigger = OnStartUpTrigger::new(0);
+        assert!(trigger.is_pre_process());
+    }
+
+    #[test]
+    fn trigger() {
+        let file = tempfile::tempdir().unwrap();
+        let logfile = LogFile {
+            writer: &mut None,
+            path: file.path(),
+            len: 0,
+        };
+
+        let trigger = OnStartUpTrigger::new(1);
+        let result = trigger.trigger(&logfile).unwrap();
+        assert_eq!(result, false); // It should not be triggered with the min_size 1.
+
+        let trigger = OnStartUpTrigger::new(0);
+        let result = trigger.trigger(&logfile).unwrap();
+        assert_eq!(result, true); // First time, it should be triggered.
+
+        let result = trigger.trigger(&logfile).unwrap();
+        assert_eq!(result, false); // Second time, it should not be triggered.
+    }
```

Comment:
I'd like to see all unique combinations of states tested.
What about when the file size < min?
What about file size <= min and not a startup call?


The following table is what is currently tested, and you can just fill it out similar to a truth table.

| case | File size | Min size | Startup |  
|------|-----------|----------|---------|
|      | 0         | 0        | true    |  
|      | 0         | 1        | true    |  
|      | 0         | 1        | false   | 

optional suggestion: I would prefer to see the comment about what is tested on a separate line just before the test case instead of at the assert point.

Reply:
Code updates look good. Calling `call_once` directly was the perfect fix for that possible race condition. Doesn't hurt to throw in a default test as well. Once we get the updated tests in I'm happy.